deep_quoridor/src/quoridor_env.py

```diff
@@ -80,22 +90,206 @@ def step(self, action):
+            if self._is_wall_between(old_row, old_col, new_row, new_col):
+                continue
+
+            # Check if the opponent is in the target position
+            if self.positions[self._opponent(agent_id)] == (new_row, new_col):
+                # Can we jump straight over the opponent?
+                straight_row = new_row + row
+                straight_col = new_col + col
+
+                if (
+                    # That new position falls within the board
+                    self._is_in_board(straight_row, straight_col)
+                    # That new position is not blocked by a wall behind them
+                    and not self._is_wall_between(new_row, new_col, straight_row, straight_col)
+                ):
+                    mask[self.action_params_to_index(straight_row, straight_col, 0)] = 1
+                    continue
+                else:
+                    # Try the diagonals, to the sides of the oponent
+                    for side in [1, -1]:
+                        if row == 0:
+                            diag_row = new_row + side
+                            diag_col = new_col
+                        else:
+                            diag_row = new_row
+                            diag_col = new_col + side
+
+                        if (
+                            # That new position falls within the board
+                            self._is_in_board(diag_row, diag_col)
+                            # That new position is not blocked by a wall
+                            and not self._is_wall_between(new_row, new_col, diag_row, diag_col)
+                        ):
+                            mask[self.action_params_to_index(diag_row, diag_col, 0)] = 1
+                    continue
+
+            # If we get to this point it's because the target is within the board,
+            # not blocked by a wall and the opponent is not on the way
+            mask[self.action_params_to_index(new_row, new_col, 0)] = 1
+
+        # TODO: Valid wall placements
+
+        return mask
+
+    def _get_info(self):
+        # This is for now unused, returning empty dict
+        return {}
+
+    def render(self):
+        board_str = ""
+
+        for row in range(self.board_size):
+            # Render board row with player positions and vertical walls
+            for col in range(self.board_size):
+                if self.positions["player_0"] == (row, col):
+                    board_str += " P0 "
+                elif self.positions["player_1"] == (row, col):
+                    board_str += " P1 "
+                else:
+                    board_str += " .  "
+
+                # Render vertical walls (|)
+                if col < self.wall_size:
+                    if row < self.wall_size and self.walls[row, col, 0]:
+                        board_str += "|"
+                    elif row > -1 and self.walls[row - 1, col, 0]:  # Continuation of a vertical wall
```

Comment:
I'm not sure why you changed to `row > -1` from `row > 0`, but I think it should be `row > 0`, otherwise if it's 0, we're getting the last row.  E.g. I run into this when testing dfs, I was placing a vertical wall on the bottom and it was overflowing to the top because of this

Reply:
I agree. I don't know why this ended up changing. I don't even remember adding the check in the first place!
I agree it needs to be row > 0 but it's for a very subtle reason! It's because otherwise row - 1 can evaluate to -1, which in python means "the last one in the array", not what we want! 
I'm fixing it now.